ahnlich/ai/src/engine/ai/providers/fastembed.rs

```diff
@@ -0,0 +1,71 @@
+use fastembed::{EmbeddingModel, InitOptions, ImageEmbeddingModel, TextEmbedding};
+use hf_hub::{api::sync::ApiBuilder, Cache};
+use ahnlich_types::ai::AIModel;
+use std::convert::TryFrom;
+use std::path::PathBuf;
+use crate::engine::ai::models::Model;
+
+
+pub struct FastEmbedProvider {
+    cache_location: PathBuf,
+}
+
+enum FastEmbedModel {
+    Text(EmbeddingModel),
+    Image(ImageEmbeddingModel),
+}
+
+#[derive(Debug)]
+pub enum Error {
+    ModelNotFound { reason: String},
+}
+
+impl TryFrom<&AIModel> for FastEmbedModel {
+    type Error = Error;
+
+    fn try_from(model: &AIModel) -> Result<Self, Self::Error> {
+        let model_modality: Model = model.into();
+        let model_type = match model_modality {
+            Model::Text { .. } => {
+                let model_type = match model {
+                    AIModel::AllMiniLML6V2 => EmbeddingModel::AllMiniLML6V2,
+                    AIModel::AllMiniLML12V2 => EmbeddingModel::AllMiniLML12V2,
+                    AIModel::BGEBaseEnV15 => EmbeddingModel::BGEBaseENV15,
+                    AIModel::BGELargeEnV15 => EmbeddingModel::BGELargeENV15,
+                    _ => return Err(Error::ModelNotFound {reason: "Model not supported".to_string()})
+                };
+                FastEmbedModel::Text(model_type)
+            },
+            Model::Image { .. } => {
+                let model_type = match model {
+                    AIModel::Resnet50 => ImageEmbeddingModel::Resnet50,
+                    AIModel::ClipVitB32 => ImageEmbeddingModel::ClipVitB32,
+                    _ => return Err(Error::ModelNotFound {reason: "Model not supported".to_string()})
+                };
+                FastEmbedModel::Image(model_type)
+            }
+        };
+        Ok(model_type)
+    }
+}
+
+
+impl FastEmbedProvider {
+    pub fn new(cache_location: PathBuf) -> Self {
+        FastEmbedProvider {
+            cache_location: cache_location
+        }
+    }
+
+    pub fn download(&self, model: &AIModel) {
+        let model_type = FastEmbedModel::try_from(model).expect("Model not found");
+
+        if let FastEmbedModel::Text(model_type) = model_type {
+            let cache = Cache::new(self.cache_location.clone());
```

Comment:
Do we need to do any checks if the model already exists in cache and confirm using the hash? Or does `Cache already handle that internally?

Reply:
They do checks to see if the model exists here:

https://github.com/huggingface/hf-hub/blob/a66bdcf2f2664a55ab3a1415184c7aebd8ac4ed5/src/api/sync.rs#L472-L478

My concern is that that check does not look like any kind of hashing involved, from looking deeper, there are hashed values but it just looks like checking if the file exists (the file is named by the hash value).

Reply:
Then I think we should use the checksums for ensuring correctness, else I predict weird things in our future

Reply:
To be left for a different PR